File: python_server/gaoranger.py
class RAD():
    def __init__(self):
        self.init_states()
        self.actions = []
        self.choreography = []
        self.PROBE_SWITCH = {
            "states": self.probe_states,
            "actions": self.probe_actions,
            "choreography": self.probe_choreography
        }
        self.MOTOR = [
            "base",
            "shoulder",
            "elbow",
            "wrist",
            "rotate",
            "gripper"
        ]
        self.LIMITS = {
            "base": (0, 180),
            "shoulder": (15, 165),
            "elbow": (0, 180),
            "wrist": (0, 180),
            "rotate": (0, 180),
            "gripper": (10, 73)
        }
        
    def init_states(self):
        self.states = {
            "led": False,
            "base": 0,
            "shoulder": 40,
            "elbow": 180,
            "wrist": 0,
            "rotate": 170,
            "gripper": 73
        }

    def probe(self, part):
        return self.PROBE_SWITCH[part]()

    def probe_states(self):
        return self.states

    def probe_actions(self):
        return self.actions

    def probe_choreography(self):
        return self.choreography

    def action(self, act, param):
        if type(act) == int:
            act = self.MOTOR[act]
        if act == "led":
            return self.act_led(param)
        else:
            return self.act_motor(act, param)

    def act_led(self, arg):
        if arg == "toggle":
            self.states["led"] = not self.states["led"]
        if arg == "on":
            self.states["led"] = True
        if arg == "off":
            self.states["led"] = False
        return True
        
    def act_motor(self, act, arg):
        arg = int(arg)
        target = self.states[act] + arg
        target = self.limit_crip(act, target)
        self.states[act] = target
        return True
    
    def limit_crip(self, motor, val):
        ll, ul = self.LIMITS[motor]
        if val < ll: return ll
        if val > ul: return ul
        return val
        
    def reset(self):
        self.init_states()
        return self.states
        
    def set_states(self, angles):
        for i, angle in enumerate(angles):
            angle = int(angle)
            motor = self.MOTOR[i]
            angle = self.limit_crip(motor, angle)
            self.states[motor] = angle
        return True
        
    def record_action(self, inst, *param):
        if inst == "reset":
            self.actions.append({"inst": "reset"})
            return
        if inst == "action":
            self.actions.append({"inst": "action", "actor": param[0], "param": param[1]})
            return
        if inst == "toward":
            self.actions.append({"inst": "toward", "angles": param[0]})
            return
    
    def record_choreo(self, actions):
        self.choreography.append(actions)
            

File: arduino_server/libraries/Gaoranger/Gaoranger.cpp
#include "Gaoranger.h"

STATES::STATES(void){
}

STATES::STATES(int _base, int _shoulder, int _elbow, int _wrist_ver, int _wrist_rot, int _gripper){
	base = _base;
	shoulder = _shoulder;
	elbow = _elbow;
	wrist_ver = _wrist_ver;
	wrist_rot = _wrist_rot;
	gripper = _gripper;
}

void STATES::outjson(YunClient client){
	client.print("{\"base\":");
	client.print(base);
	client.print(",\"shoulder\":");
	client.print(shoulder);
	client.print(",\"elbow\":");
	client.print(elbow);
	client.print(",\"wrist\":");
	client.print(wrist_ver);
	client.print(",\"rotate\":");
	client.print(wrist_rot);
	client.print(",\"gripper\":");
	client.print(gripper);
	client.print("}");
}

RAD::RAD(void){
	reset();
}

STATES RAD::reset(void){
	states = *(new STATES(0, 40, 180, 0, 170, 73));
	return states;
}

STATES RAD::probe(void){
	return states;
}

STATES RAD::do_action(int action, int param){
	int target;
	switch(action){
	case 0: // 0 - 180
		target = states.base + param;
		if(target < 0){
			target = 0;
		}
		if(target > 180){
			target = 180;
		}
		states.base = target;
		break;
	case 1: // 15 - 165
		target = states.shoulder + param;
		if(target < 15){
			target = 15;
		}
		if(target > 165){
			target = 165;
		}
		states.shoulder = target;
		break;
	case 2: // 0 - 180
		target = states.elbow + param;
		if(target < 0){
			target = 0;
		}
		if(target > 180){
			target = 180;
		}
		states.elbow = target;
		break;
	case 3: // 0 - 180
		target = states.wrist_ver + param;
		if(target < 0){
			target = 0;
		}
		if(target > 180){
			target = 180;
		}
		states.wrist_ver = target;
		break;
	case 4: // 0 - 180
		target = states.wrist_rot + param;
		if(target < 0){
			target = 0;
		}
		if(target > 180){
			target = 180;
		}
		states.wrist_rot = target;
		break;
	case 5: // 10 - 73
		target = states.gripper + param;
		if(target < 10){
			target = 10;
		}
		if(target > 73){
			target = 73;
		}
		states.gripper = target;
		break;
	}
	return states;
}

STATES RAD::set_states(STATES states_){
	if(states_.base < 0){
		states_.base = 0;
	}
	if(states_.base > 180){
		states_.base = 180;
	}
	if(states_.shoulder < 15){
		states_.shoulder = 15;
	}
	if(states_.shoulder > 165){
		states_.shoulder = 165;
	}
	if(states_.elbow < 0){
		states_.elbow = 0;
	}
	if(states_.elbow > 180){
		states_.elbow = 180;
	}
	if(states_.wrist_ver < 0){
		states_.wrist_ver = 0;
	}
	if(states_.wrist_ver > 180){
		states_.wrist_ver = 180;
	}
	if(states_.wrist_rot < 0){
		states_.wrist_rot = 0;
	}
	if(states_.wrist_rot > 180){
		states_.wrist_rot = 180;
	}
	if(states_.gripper < 10){
		states_.gripper = 10;
	}
	if(states_.gripper > 73){
		states_.gripper = 73;
	}
	states = states_;
	return states_;
}

